zcash_primitives/src/extensions/transparent.rs

```diff
@@ -0,0 +1,199 @@
+/// and extension-specific types. The payload field of this struct
+/// is treated as opaque to all but extension corresponding to the
+/// encapsulated extension_id value.
+#[derive(Clone, Debug, PartialEq)]
+pub struct Precondition {
+    pub extension_id: u32,
+    pub mode: u32,
+    pub payload: Vec<u8>,
+}
+
+impl Precondition {
+    /// Produce the intermediate format for an extension-specific precondition
+    /// type.
+    pub fn from<P: ToPayload>(extension_id: u32, value: &P) -> Precondition {
+        let (mode, payload) = value.to_payload();
+        Precondition {
+            extension_id,
+            mode,
+            payload,
+        }
+    }
+
+    /// Attempt to parse an extension-specific precondition value from the
+    /// intermediate representation.
+    pub fn try_to<P: FromPayload>(&self) -> Result<P, P::Error> {
+        P::from_payload(self.mode, &self.payload)
+    }
+}
+
+/// Data that satisfies the precondition for prior encumbered funds, enabling them to be
+/// spent.
+///
+/// This struct is an intermediate representation between the
+/// serialized binary format which is used inside of a transaction
+/// and extension-specific types. The payload field of this struct
+/// is treated as opaque to all but extension corresponding to the
+/// encapsulated extension_id value.
+#[derive(Clone, Debug, PartialEq)]
+pub struct Witness {
+    pub extension_id: u32,
+    pub mode: u32,
+    pub payload: Vec<u8>,
+}
+
+impl Witness {
+    /// Produce the intermediate format for an extension-specific witness
+    /// type.
+    pub fn from<P: ToPayload>(extension_id: u32, value: &P) -> Witness {
+        let (mode, payload) = value.to_payload();
+        Witness {
+            extension_id,
+            mode,
+            payload,
+        }
+    }
+
+    /// Attempt to parse an extension-specific witness value from the
+    /// intermediate representation.
+    pub fn try_to<P: FromPayload>(&self) -> Result<P, P::Error> {
+        P::from_payload(self.mode, &self.payload)
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub enum Error<E> {
+    InvalidExtensionId(u32),
+    ProgramError(E),
+}
+
+impl<E: fmt::Display> fmt::Display for Error<E> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Error::InvalidExtensionId(extension_id) => {
+                write!(f, "Unrecognized program type id {}", extension_id)
+            }
+
+            Error::ProgramError(err) => write!(f, "Program error: {}", err),
+        }
+    }
+}
+
+/// This is the primary trait which must be implemented by an extension type
+/// for that type to be eligible for inclusion in Zcash consensus rules.
+pub trait Extension<C> {
+    /// Extension-specific precondition type. The extension will need to implement
+    /// [`FromPayload<Error = Self::Error>`] for this type in order for their
+    /// extension to be eligible for integration into consensus rules.
+    type P;
```

Comment:
I think this is too vague of a name; I don't think calling it Precondition would cause any shadowing issues, right?